### DISCUSSION TITLE
Timer workflow: Schedule to send email on [Due date] = [Today] records.

### DISCUSSION THREAD

**Discussion by orchardc** [OrchardCMS/OrchardCore#15888] [Q&A]
What is the best approach to implement the following feature?

I have a custom module which uses a custom data table, call it: Table1; In Table1 there are some columns, one of them call [Due date]. I'd like to implement a feature that for all records in Table1, if [Due Date] = today's date, then send an email (a column in Table1) accordingly; it will be a schedule task running on daily basis.

I have added the Timer module and setup a workflow, but not sure which Task can fit the above requirement. It is understood that custom code is required to scan Table1, **but try to make it configurable where possible**. Any suggestion is welcomed!

![image](https://user-images.githubusercontent.com/16193516/59829730-09983d80-9336-11e9-83d6-33f30f03f55c.png)

**Comment by zl2fxy** [comment]
they are ineffective for your work，you will add a backgroudTask to do it。
you need  code of custom task。Reference resources has timer task;

**Comment by orchardc** [comment]
Thanks @zl2fxy , with the background task, is it configurable? 

I understand the custom task will require coding (to do Table scanning); but overall, is it configurable (to some extend) just like Workflow?

**Comment by zl2fxy** [comment]
1.the background task, it is configurable,it will work with your codes.
2.i think you are wrong,you don't scan the table，It takes a lot of time and it is meaningless if  your have
a lot of unqualified datas

**Comment by orchardc** [comment]
Thanks @zl2fxy , OK I get it, will look at some reference about background tasks.

Could you point out - without scanning the Table1 to determine the DueDate records, how would you do it in this case?  Just to be clear, why I say "scan" I mean SQL query, sorry for the confusion caused.

**Comment by sebastienros** [comment]
You could implement everything with a custom background task. Look at the classes that already implement `IBackgroundTask`.

We could do that in Workflow, if we had a `GlobalMethod` provider in the SQL module to expose named  SQL queries in the javascript engine. I assume it would be a good addition. Same for Lucene queries.

**Comment by orchardc** [comment]
Thanks @sebastienros , I see the below implementation, however it looks like there is no way to pass in parameters from the configuration page ?

For example, how can I pass a INT (number) from the Admin's Task configuration page to DoWorkAsync() ?

At the moment, there is just a Schedule text box.
![image](https://user-images.githubusercontent.com/16193516/59879196-0a66b900-93a2-11e9-85d4-f94d157cebf7.png)

`        public Task DoWorkAsync(IServiceProvider serviceProvider, CancellationToken cancellationToken)
        {
            _count++;
            return Task.CompletedTask;
        }`

**Comment by zl2fxy** [comment]
> Thanks @zl2fxy , OK I get it, will look at some reference about background tasks.
> 
> Could you point out - without scanning the Table1 to determine the DueDate records, how would you do it in this case? Just to be clear, why I say "scan" I mean SQL query, sorry for the confusion caused.

IBackgroundTask is use for your work，it has  a perfect mechanism to manage time tasks.you should do it that is you triggering BackgroundTask when you storing your data。
other，you can pass parameters for your task with your customized codes，
liquid engine，javascript engine and so on

**Comment by orchardc** [comment]
Thanks @zl2fxy ,

I can see the [Set Parameter] when configuring a Workflow, this could be the Liquid engine you mentioned?  

But the Workflow is different than the Background Tasks, how to connect both together? You can see from the below picture, I have already got a Background Task but it is not listed in the Workflow editor.

In Tasks Editor;
![image](https://user-images.githubusercontent.com/16193516/59903937-33b33380-93fa-11e9-858a-8c55464929a6.png)

In Workflow Editor:

![image](https://user-images.githubusercontent.com/16193516/59903819-ddde8b80-93f9-11e9-99cc-2025b0cfb913.png)
